File: 56-merge-intervals/merge-intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
    if (intervals.length <= 1) return intervals

    intervals.sort((a, b) => a[0] - b[0])

    let writeIndex = 0

    for (let i = 1; i < intervals.length; i++) {
        if (intervals[writeIndex][1] >= intervals[i][0]) {
            intervals[writeIndex][1] = Math.max(intervals[writeIndex][1], intervals[i][1])
        } else {
            writeIndex++
            intervals[writeIndex] = intervals[i]
        }
    }

    console.log(intervals)
    console.log(writeIndex)
    intervals.length = writeIndex + 1
    return intervals
};

File: 56-merge-intervals/merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if not intervals:
            return []

        n = len(intervals)
        intervals.sort(key=lambda x: x[0])
        merged = [intervals[0]]

        for i in range(1, n):
            prev_start, prev_end = merged[-1]
            curr_start, curr_end = intervals[i]

            if prev_end >= curr_start:
                end = max(prev_end, curr_end)
                merged[-1][1] = end
            else:
                merged.append(intervals[i])

        return merged
